File: 1034_color_border/color_border.cpp
/*
 * @Date: 2021-12-07 00:56:23
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-07 01:19:25
 */

#include <cassert>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

typedef pair<int, int> pii;

class Solution {
 public:
  vector<vector<int>> colorBorder_bfs(vector<vector<int>>& grid, int row,
                                      int col, int color) {
    int m = grid.size(), n = grid[0].size();
    vector<vector<bool>> visited(m, vector<bool>(n, false));
    vector<pii> borders;
    int originalColor = grid[row][col];
    int direc[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    queue<pii> q;
    q.emplace(row, col);
    visited[row][col] = true;
    while (!q.empty()) {
      pii& node = q.front();
      q.pop();
      int x = node.first, y = node.second;

      bool isBorder = false;
      for (int i = 0; i < 4; i++) {
        int nx = direc[i][0] + x, ny = direc[i][1] + y;
        if (!(nx >= 0 && nx < m && ny >= 0 && ny < n &&
              grid[nx][ny] == originalColor)) {
          isBorder = true;
        } else if (!visited[nx][ny]) {
          visited[nx][ny] = true;
          q.emplace(nx, ny);
        }
      }
      if (isBorder) {
        borders.emplace_back(x, y);
      }
    }
    for (auto& [x, y] : borders) {
      grid[x][y] = color;
    }
    return grid;
  }

  vector<vector<int>> colorBorder_dfs(vector<vector<int>>& grid, int row,
                                      int col, int color) {
    int m = grid.size(), n = grid[0].size();
    vector<vector<bool>> visited(m, vector<bool>(n, false));
    vector<pii> borders;
    int originalColor = grid[row][col];
    visited[row][col] = true;
    dfs(grid, row, col, visited, borders, originalColor);
    for (auto& [x, y] : borders) {
      grid[x][y] = color;
    }
    return grid;
  }

  void dfs(vector<vector<int>>& grid, int x, int y,
           vector<vector<bool>>& visited, vector<pii>& borders,
           int originalColor) {
    int m = grid.size(), n = grid[0].size();
    bool isBorder = false;
    int direc[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    for (int i = 0; i < 4; i++) {
      int nx = direc[i][0] + x, ny = direc[i][1] + y;
      if (!(nx >= 0 && nx < m && ny >= 0 && ny < n &&
            grid[nx][ny] == originalColor)) {
        isBorder = true;
      } else if (!visited[nx][ny]) {
        visited[nx][ny] = true;
        dfs(grid, nx, ny, visited, borders, originalColor);
      }
    }
    if (isBorder) {
      borders.emplace_back(x, y);
    }
  }
};

int main() {
  {
    vector<vector<int>> grid = {{1, 1}, {1, 2}};
    assert((Solution().colorBorder_bfs(grid, 0, 0, 3) ==
            vector<vector<int>>{{3, 3}, {3, 2}}));
  }
  {
    vector<vector<int>> grid = {{1, 1}, {1, 2}};
    assert((Solution().colorBorder_dfs(grid, 0, 0, 3) ==
            vector<vector<int>>{{3, 3}, {3, 2}}));
  }
  {
    vector<vector<int>> grid = {{1, 2, 2}, {2, 3, 2}};
    assert((Solution().colorBorder_bfs(grid, 0, 1, 3) ==
            vector<vector<int>>{{1, 3, 3}, {2, 3, 3}}));
  }
  {
    vector<vector<int>> grid = {{1, 2, 2}, {2, 3, 2}};
    assert((Solution().colorBorder_dfs(grid, 0, 1, 3) ==
            vector<vector<int>>{{1, 3, 3}, {2, 3, 3}}));
  }
  {
    vector<vector<int>> grid = {{1, 1, 1}, {1, 1, 1}, {1, 1, 1}};
    assert((Solution().colorBorder_bfs(grid, 1, 1, 2) ==
            vector<vector<int>>{{2, 2, 2}, {2, 1, 2}, {2, 2, 2}}));
  }
  {
    vector<vector<int>> grid = {{1, 1, 1}, {1, 1, 1}, {1, 1, 1}};
    assert((Solution().colorBorder_dfs(grid, 1, 1, 2) ==
            vector<vector<int>>{{2, 2, 2}, {2, 1, 2}, {2, 2, 2}}));
  }
}

File: 1034_color_border/color_border.go
/*
 * @Date: 2021-12-07 00:56:37
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-07 01:17:21
 */

package main

import "reflect"

var dirs = []struct{ x, y int }{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

func colorBorder_dfs(grid [][]int, row, col, color int) [][]int {
	m, n := len(grid), len(grid[0])
	type point struct{ x, y int }
	borders := []point{}
	originalColor := grid[row][col]
	vis := make([][]bool, m)
	for i := range vis {
		vis[i] = make([]bool, n)
	}

	var dfs func(int, int)
	dfs = func(x, y int) {
		vis[x][y] = true
		isBorder := false
		for _, dir := range dirs {
			nx, ny := x+dir.x, y+dir.y
			if !(0 <= nx && nx < m && 0 <= ny && ny < n && grid[nx][ny] == originalColor) {
				isBorder = true
			} else if !vis[nx][ny] {
				vis[nx][ny] = true
				dfs(nx, ny)
			}
		}
		if isBorder {
			borders = append(borders, point{x, y})
		}
	}
	dfs(row, col)

	for _, p := range borders {
		grid[p.x][p.y] = color
	}
	return grid
}

func colorBorder_bfs(grid [][]int, row, col, color int) [][]int {
	m, n := len(grid), len(grid[0])
	type point struct{ x, y int }
	borders := []point{}
	originalColor := grid[row][col]
	vis := make([][]bool, m)
	for i := range vis {
		vis[i] = make([]bool, n)
	}

	q := []point{{row, col}}
	vis[row][col] = true
	for len(q) > 0 {
		p := q[0]
		q = q[1:]
		x, y := p.x, p.y
		isBorder := false
		for _, dir := range dirs {
			nx, ny := x+dir.x, y+dir.y
			if !(0 <= nx && nx < m && 0 <= ny && ny < n && grid[nx][ny] == originalColor) {
				isBorder = true
			} else if !vis[nx][ny] {
				vis[nx][ny] = true
				q = append(q, point{nx, ny})
			}
		}
		if isBorder {
			borders = append(borders, point{x, y})
		}
	}

	for _, p := range borders {
		grid[p.x][p.y] = color
	}
	return grid
}

func main() {
	assert := func(a, b [][]int) {
		if !reflect.DeepEqual(a, b) {
			panic("Not Passed")
		}
	}
	assert(colorBorder_dfs([][]int{{1, 1}, {1, 2}}, 0, 0, 3), [][]int{{3, 3}, {3, 2}})
	assert(colorBorder_bfs([][]int{{1, 1}, {1, 2}}, 0, 0, 3), [][]int{{3, 3}, {3, 2}})
	assert(colorBorder_dfs([][]int{{1, 2, 2}, {2, 3, 2}}, 0, 1, 3), [][]int{{1, 3, 3}, {2, 3, 3}})
	assert(colorBorder_bfs([][]int{{1, 2, 2}, {2, 3, 2}}, 0, 1, 3), [][]int{{1, 3, 3}, {2, 3, 3}})
	assert(colorBorder_dfs([][]int{{1, 1, 1}, {1, 1, 1}, {1, 1, 1}}, 1, 1, 2), [][]int{{2, 2, 2}, {2, 1, 2}, {2, 2, 2}})
	assert(colorBorder_bfs([][]int{{1, 1, 1}, {1, 1, 1}, {1, 1, 1}}, 1, 1, 2), [][]int{{2, 2, 2}, {2, 1, 2}, {2, 2, 2}})
}
